File: _stats_build/statistics/wr_worst_counting.rb
require_relative "abstract/wr_round_history"

class WrWorstCounting < WrRoundHistory
  def initialize
    @title = "Worst counting solve"
    @title_zh = "最差有效单次"
    @note = "Worst counting solve: the worst single that counts into the Ao5 (excluding the dropped best and worst)."
    @note_zh = "最差有效单次：计入 Ao5 的最差单次（排除去掉的最好和最差成绩）。"
    @table_header = { "Result" => :right, "Improvement" => :right, "Days" => :right, "Person" => :left, "Competition" => :left, "Date" => :left, "Details" => :left }
  end

  # NOTE: Worst Counting = Ao5 中计入平均的3次成绩中的最差一次
  # 即去掉最好和最差后，剩余3次中的最大值
  # 等价于排序后第4个值（0-indexed: sorted[3]）
  def compute_metric(values, r)
    valid = values.select { |v| v > 0 }
    invalid_count = values.count { |v| v <= 0 }

    case invalid_count
    when 0
      # 5个全部有效，去掉最好和最差，取剩余3个中最大的
      sorted = valid.sort
      sorted[3]  # 第4小 = counting 中的最差
    when 1
      # 1个 DNF：DNF 是最差，去掉后从4个中去掉最好，取剩余3个中最大的
      sorted = valid.sort
      sorted[2]  # 4个有效中第3大 = counting 中的最差
    else
      nil  # 2个及以上 DNF，Ao5 本身就是 DNF
    end
  end
end

File: trainer/packages/stats-build/src/statistics/wr_worst_counting.ts
// NOTE: Worst Counting Solve——Ao5 中计入平均的 3 次中的最差一次
// 与 Ruby _stats_build/statistics/wr_worst_counting.rb 1:1 对应
import { RoundMetric } from '../core/round_metric.js';
import { EVENTS_WITH_AO5 } from '../core/events.js';

export class WrWorstCounting extends RoundMetric {
  constructor() {
    super();
    this.title = 'Worst counting solve';
    this.titleZh = '最差有效';
    this.note = 'Worst counting solve: the worst single that counts into the Ao5 (excluding the dropped best and worst).';
    this.noteZh = '最差有效：Ao5 去掉最好和最差后，剩余 3 次中的最大值。';
    this.tableHeader = {
      'Result': 'right', 'Improvement': 'right', 'Days': 'right',
      'Person': 'left', 'Date': 'left', 'Competition': 'left', 'Details': 'left',
    };
  }

  targetEvents() { return EVENTS_WITH_AO5; }

  // NOTE: Worst Counting = 去掉最好和最差后的最大值
  computeMetric(values: number[]): number | null {
    const valid = values.filter(v => v > 0).sort((a, b) => a - b);
    const invalidCount = values.filter(v => v <= 0).length;
    switch (invalidCount) {
      case 0: return valid[3] ?? null;  // 5 全有效，第 4 小 = counting 最差
      case 1: return valid[2] ?? null;  // 1 DNF，4 有效中第 3 个
      default: return null;
    }
  }
}
